mat64/product.go

```diff
@@ -0,0 +1,190 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat64
+
+import "fmt"
+
+// Product calculates the product of the given factors and places the result in
+// the receiver. The order of multiplication operations is optimized to minimize
```

Comment:
I don't really want this to become that complicated. I think there is room elsewhere in gonum to do that kind of thing (for example the thing @jonlawlor is working on).